enhancements/update/clusterversion-history-pruning.md

```diff
@@ -0,0 +1,166 @@
+But certain history entries are more helpful than others such as the cluster's initial version, first minor update, and completed rather than partial version updates.
+Long lived clusters can have a history that contain very little of the aforementioned helpful entries.
+
+### User Stories
+
+* As a OCP product team member, I want a cluster's ClusterVersion history to retain entries deemed to be the most helpful in identifying and resolving cluster issues.
+
+* As a cluster administrator, I want a cluster's ClusterVersion history to retain entries deemed to be the most helpful in identifying and resolving cluster issues.
+
+* As a cluster administrator, I want a cluster's ClusterVersion history to cover the cluster's entire lifespan.
+
+### Goals
+
+* In generally descending priority, history pruning algorithm will prefer:
+  * initial versions and final version
+  * complete over partially installed versions
+  * first and last minor versions
+  * more recent entries
+* Preserve useful version history rather than simply preserving more version history.
+* Structure pruning algorithm to allow easy tuning/tweaking.
+* Implementation will require no API changes.
+* Pruning will be accomplished with a single pass through the data.
+
+### Non-Goals
+
+* ClusterVersion history will contain no explicit indication that it has been pruned however pruning details will be logged.
+
+## Proposal
+
+### Workflow Description
+
+When a new entry is to be added to the history queue and the queue contains 100 entries, the new entry and each existing history entry will be ranked:
+```go
+rank =
+  1000 * (isTheInitialEntry or isAFinalEntry or isTheMostRecentCompletedEntry)
+  + 30 * (isTheFirstCompletedInAMinor or isTheLastCompletedInAMinor)
+  + 20 * (isPartialPortionOfMinorTransition)
+  - 20 * (isPartialWithinAZStream)
+  - 1.01 * sliceIndex
+```
+The lowest ranked entry will be removed, or, in the case of the new entry not added, maintaining the queue size of 100 entries.
+The removal including details of why an entry was chosen for removal will be logged.
+
+The weights used to rank a given entry are based on the following criteria:
+* `1000` - considered most important to keep
+* `30` - interesting but not critical  
+* `20` - partial stops during a minor transition are interesting, but not as interesting as the completed versions that bookend the minor transition
+* `-20` - partials between completed releases within the same minor, not very useful
+* `-1.01` - prefer more recent entries, avoid ties
+
+`isAFinalEntry` will be true for the 5 most recent history entries.
+
+Prototyping against ClusterVersion histories from a variety of actual customer clusters will help determine the correct criteria and corresponding weights that result in the most useful history.
+
+godocs will be created explaining that ClusterVersion history may be missing entries (has been pruned).
```

Comment:
Turns out [`history` godocs][1] have had this since 2019:

> Only a limited amount of update history is preserved.

So I don't think we need to change the godocs.  It is worth pointing out these existing godocs in the enhancement.

[1]: https://github.com/openshift/api/blame/63b58097a95d1b053cf038f56f89afc42524f598/config/v1/types_cluster_version.go#L103-L104